cli/tools/vendor/mappings.rs

```diff
@@ -0,0 +1,286 @@
+      mappings.insert(root, base_dir);
+    }
+
+    // resolve all the "proxy" paths to use for when an x-typescript-types header is specified
+    for module in remote_modules {
+      if let Some((
+        _,
+        Resolved::Ok {
+          specifier, range, ..
+        },
+      )) = &module.maybe_types_dependency
+      {
+        // hack to tell if it's a x-typescript-types header
+        let is_ts_types_header =
+          range.start == Position::zeroed() && range.end == Position::zeroed();
+        if is_ts_types_header {
+          let module_path = mappings.get(&module.specifier).unwrap();
+          let proxied_path = get_unique_path(
+            path_with_stem_suffix(module_path, ".proxied"),
+            &mut mapped_paths,
+          );
+          proxies.insert(
+            module.specifier.clone(),
+            ProxiedModule {
+              output_path: proxied_path,
+              declaration_specifier: specifier.clone(),
+            },
+          );
+        }
+      }
+    }
+
+    Ok(Self {
+      output_dir: ModuleSpecifier::from_directory_path(output_dir).unwrap(),
+      mappings,
+      base_specifiers,
+      proxies,
+    })
+  }
+
+  pub fn output_dir(&self) -> &ModuleSpecifier {
+    &self.output_dir
+  }
+
+  pub fn local_uri(&self, specifier: &ModuleSpecifier) -> ModuleSpecifier {
+    if specifier.scheme() == "file" {
+      specifier.clone()
+    } else {
+      let local_path = self.local_path(specifier);
+      if specifier.path().ends_with('/') {
+        ModuleSpecifier::from_directory_path(&local_path)
+      } else {
+        ModuleSpecifier::from_file_path(&local_path)
+      }
+      .unwrap_or_else(|_| {
+        panic!("Could not convert {} to uri.", local_path.display())
+      })
+    }
+  }
+
+  pub fn local_path(&self, specifier: &ModuleSpecifier) -> PathBuf {
+    if specifier.scheme() == "file" {
+      specifier.to_file_path().unwrap()
+    } else {
+      self
+        .mappings
+        .get(specifier)
+        .as_ref()
+        .unwrap_or_else(|| {
+          panic!("Could not find local path for {}", specifier)
+        })
+        .to_path_buf()
+    }
+  }
+
+  pub fn relative_path(
+    &self,
+    from: &ModuleSpecifier,
+    to: &ModuleSpecifier,
+  ) -> String {
+    let mut from = self.local_uri(from);
+    let to = self.local_uri(to);
+
+    // workaround using parent directory until https://github.com/servo/rust-url/pull/754 is merged
+    if !from.path().ends_with('/') {
+      let local_path = self.local_path(&from);
+      from = ModuleSpecifier::from_directory_path(local_path.parent().unwrap())
+        .unwrap();
+    }
+
+    // workaround for url crate not adding a trailing slash for a directory
+    // it seems to be fixed once a version greater than 2.2.2 is released
+    let is_dir = to.path().ends_with('/');
+    let mut text = from.make_relative(&to).unwrap();
+    if is_dir && !text.ends_with('/') && to.query().is_none() {
+      text.push('/');
+    }
```

Comment:
It's a bit of a pain to test this struct because it involves creating a module graph. I thought about extracting some of the functionality in here out, but it adds a bit of indirection. The logic is temporary workarounds and the code in here is tested transitively—if I delete either of the if statements or start flipping conditions then the vendor tests will fail.